### ISSUE TITLE
[Bug]: `keyboard.press('Tab')` does not set active element

### ISSUE THREAD

**Issue by JakobJingleheimer** [puppeteer/puppeteer#9528]
### Bug expectation

I expect pressing the `Tab` key to behave as it does in the browser: focus the next focusable element in the document.

Instead, it does not set an active element.

### Bug behavior

- [ ] Flaky
- [ ] PDF

### Minimal, reproducible example

```mjs
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch();

const page = await browser.newPage();

await page.setContent('<input id="login">');

await page.waitForSelector('#login');

await page.keyboard.press('Tab');

const id = await page.evaluate(() => document.activeElement?.id);

if (id !== 'login') {
    throw new Error('login is not active')
}

await browser.close();
```

### Error string

login is not active

### Puppeteer configuration

```TypeScript
```

### Puppeteer version

19.4.1

### Node version

18.12.0

### Package manager

npm

### Package manager version

8.19.2

### Operating system

macOS

**Comment by OrKoN** [comment]
The code is missing an await and the server part and I am not able to repro if I try to reproduce: 

```
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch();

const page = await browser.newPage();

await page.setContent('<input id="login">');

await page.waitForSelector('#login');

await page.keyboard.press('Tab');

const activeElement = await page.evaluateHandle(() => document.activeElement);

console.log({ activeElement }); // { activeElement: ElementHandle {} }
```

**Comment by JakobJingleheimer** [comment]
Sorry, I added the extra browser & page setup manually because your bot yelled at me. There is actually an await in my setup.

Hmmm. I copy-paste your code, run it, and now I get an error about id of undefined (citing the line of `document.activeElement.id`).

**Comment by OrKoN** [comment]
that's strange. Have you tried using the latest version of Puppeteer?

**Comment by OrKoN** [comment]
It does not looks like it reproduces on the bots as well. Locally, I also get no error.

**Comment by JakobJingleheimer** [comment]
I updated to 19.5.2, but no difference.

But. I think I found the source of this baffling behaviour:

> If the function passed to the `page.evaluate` returns a non-[Serializable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description) value, then `page.evaluate` resolves to `undefined`.

I don't see this in the puppeteer docs [directly](https://pptr.dev/next/api/puppeteer.page.evaluatehandle)¹, but it does appear in devdocs.io's [copy](https://devdocs.io/puppeteer/index#pageevaluatepagefunction-args). I don't know if theirs is outdated, but it's the only thing I've found to explain this behaviour.

`page.evaluate` does not like returning `document.activeElement` itself, but properties from it are okay (as is a plain object of properties):

ok:
```js
const id = await page.evaluate(() => document.activeElement.id);
```

also ok:
```js
const activeElement = await page.evaluate(() => {
  const { activeElement } = document;

  return {
    id: activeElement.id,
    type: activeElement.type,
  };
});
```

silently fails:
```js
const activeElement = await page.evaluate(() => document.activeElement);
```

Possibly related/contributing to `page.evaluate()` silently failing: in my document, there are multiple focusable elements. The first several (in terms of DOM order) are input fields; there are also some link elements (which occur after the inputs in DOM order). When the page is opened by a human in a browser, (as the first interaction with the page) pressing <kbd>tab</kbd> correctly focuses the first input field. Puppeteer, however, is focusing a seemingly random one of the links (like the 7th link of 8). As an experiment, I tried setting `tabindex`s on the inputs, but puppeteer does not behave correctly / does not respect the tabindex: the links are all focused first and then the inputs.

I'm trying to simulate a user navigating the page via keyboard, so I'm using `page.evaluate(() => document.activeElement)` to get the element that became focused after `page.keyboard.press('Tab')`.

---

¹ side-note, the puppeteer docs for `page.evaluate` have some copy-pasta: 
<img width="305" alt="docs citing wrong method" src="https://user-images.githubusercontent.com/3012099/212981356-bb543a7e-3dc8-450b-8951-181228286030.png">

**Comment by OrKoN** [comment]
@JakobJingleheimer have you think this guide that we have https://pptr.dev/guides/evaluate-javascript#return-types ?

You can get the reference to the active element with evaluateHandle, that still works.

**Comment by OrKoN** [comment]
@JakobJingleheimer how do you check it's undefined? it is defined for me, it's a JSHandle instance. Full code:

```
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch();

const page = await browser.newPage();

await page.setContent('<input id="login">');

await page.waitForSelector('#login');

await page.keyboard.press('Tab');

const output = await page.evaluateHandle(() => {
    const { activeElement } = document;
  
    return {
        activeElement,
    };
});

console.log(output)

await browser.close(); 
```

Output: 
```
❯ node focus.mjs
JSHandle {}
```

**Comment by OrKoN** [comment]
This `console.log(output) // { activeElement: undefined }` cannot be true because evaluateHandle always returns a handle so it does not know what properties exist on the object it references, i.e., you cannot use it like `output. activeElement` in the script context.

**Comment by JakobJingleheimer** [comment]
Ahh, sorry I was overly concise. I was actually checking

```js
const { activeElement } = await page.…

console.log({ activeElement })
```

I've finally been able to get the data I need with a combination of `page.evaluateHandle()` and `elementHandle.evaluate()`:

```js
const activeElement = await page.evaluateHandle(() => document.activeElement);

const {
  name,
  tagName,
  type,
} = await activeElement.evaluate(({ name, tagName, type }) => ({ name, tagName, type }));
```

However, `page.keyboard.press('Tab')` still focuses the wrong element (it's picking a random link, instead of the first focusable element).

**Comment by JakobJingleheimer** [comment]
> @JakobJingleheimer have you seen this guide that we have https://pptr.dev/guides/evaluate-javascript#return-types ?

Sorry, I missed this previously. I had not seen it. The guide's title is rather generic, so I never would have looked there for this (or really anything). Perhaps it would be good to link to it in the relevant API docs, as well as change the title to something more specific, like "Evaluating in the browser" (it's all javascript, so no need to mention that by name)?

**Comment by OrKoN** [comment]
@JakobJingleheimer could you update the ticket with the actual code that demonstrates that the active element is wrong?

**Comment by JakobJingleheimer** [comment]
I tried to create a minimal reproduction with the same fields and links, but it doesn't repro with just HTML.

The actual document is a react app, and the page in question is re-renderer as data loads (for instance, input's datalists and the links are populated based on fetch calls).

At first, I thought it was a cart-before-the-horse issue, where the links finish before the form, so I tried a more localised selector (for `page.waitForSelector()`), but that didn't make a difference.

So then I started to loop through all focusable elements, and the list is completely disorganised—something like (printing `activeElement.tagName`):

```
A
A
A
A
A
INPUT
INPUT
INPUT
BODY
BUTTON
A
A
BUTTON
INPUT
INPUT
A
…
```

The actual layout of the DOM is like this:

```
INPUT (9x)
BUTTON (2x)
A (6x)
… (various others, not together)
```

Because the browser handles this correctly but Puppeteer does not, that makes me think Puppeteer is not handling the updates/re-renders correctly.

**Comment by OrKoN** [comment]
@JakobJingleheimer hm it looks like it might due to vdom. React might update the same element with a new content but Puppeteer always holds a reference to an element so it will fetch the latest data on an element. Also, React has its own scheduler and you probably want to wait before React finished rendering before you try to grab the elements. 

You could try to build out a minimal app in React to demonstrate it but I think there is no bug here and just a race condition between React and Puppeteer.

**Comment by JakobJingleheimer** [comment]
How does one get Puppeteer to wait for react to finish rendering? I looked for that specifically, but I haven't found anything.

I thought of using `componentDidMount` to emit an event for Puppeteer to listen for, but I'm not sure it can do that.

**Comment by OrKoN** [comment]
@JakobJingleheimer at first, perhaps you could add a timeout `await new Promise(resolve => setTimeout(resolve, 3000))` before and after sending the Tab call. If you get consistent result, it would confirm it's a race with React.

**Comment by JakobJingleheimer** [comment]
Just tried that timeout, and no change to the sequence of focusable elements. I also tried setting the timeout after the page.waitForSelector checking the page has loaded (also no change).

**Comment by OrKoN** [comment]
@JakobJingleheimer sorry, cannot help more without a repro.

**Comment by JakobJingleheimer** [comment]
Mm, I understand.

I'll try to get something together.

Thanks for your help thus-far! I'm at least in a state where I can work around the disorganised focusable elements 🙂

**Comment by JakobJingleheimer** [comment]
I'm working on the repro, and I think I've got all the bits except for one final piece: When using <kbd>Tab</kbd> in puppeteer to focus on the date field (which looks like `dd/mm/yyyy 📅`), the `yyyy` gets selected; when a human focuses on it the same way (<kbd>Tab</kbd>), the `dd` gets selected.

I believe the yyyy-selection (instead of dd-selection) in puppeteer is causing the interference for typing the value (when a human does it, the value is typed from left to right across the whole input field, automatically jumping to the next date piece). With puppeteer, it types right to left within the year (first the day in the two right-most digits, then the month, pushing the day left, and then overwrites the whole thing with the year).

Can you think of anything that could cause this behaviour? My locale is `en-GB` and AFAIK, I don't have any right-to-left things set. I had `flex-direction: row-reverse` set so the picker icon would be on the left, but I disabled that to seemingly no difference.

**Comment by OrKoN** [comment]
If you click on the element like date, Puppeteer will click in the middle by default which is likely to be the mm part. The next tab then will go the yyyy part. Perhaps, it is what's happening here.

**Comment by JakobJingleheimer** [comment]
I'm pretty sure that's not it: I captured a screen recording, and when viewing it frame-by-frame, it appears the yyyy is immediately selected as soon as the field is focused.

**Comment by JakobJingleheimer** [comment]
We found the cause of both issues (random/chaotic field focus and wrong sub-field focus within date/datetime fields): React's [reconciliation](https://legacy.reactjs.org/docs/reconciliation.html).

TLDR: The workaround to set a tiny delay for Puppeteer whenever the DOM will be updated and Puppeteer will interact with that update.

Reconciliation is async process that happens at the very end of React's rendering phase, as a sort of clean-up step; it's very fast (well fast enough to finish before a human can do anything), but Puppeteer is faster—effectively jumping the gun (in Puppeteer's defence—there is no gun, just "ready, set, __").

Trying to come up with a simple reproduction was doomed from the start: It does not re-pro in a simple example because the issue requires a lot of DOM updates in order to cause reconciliation to take longer than Puppeteer.

I think an ideal solution would be to expose some way to inform Puppeteer things are "go". This is possible to know via react's [`componentDidMount`](https://react.dev/reference/react/Component#componentdidmount) and [`componentDidUpdate`](https://react.dev/reference/react/Component#componentdidupdate) (react renders from leaves to root, so a "page" as a class component would call these after all its content has finished). Would it be possible (and not excessively difficult) to broadcast an event for which Puppeteer would listen and then "go"?

This would also avoid needing `waitForSelector` (which continuously probes for a match on the provided selector).

**Comment by OrKoN** [comment]
@JakobJingleheimer you can use `page.exposeFunction` to expose a callback to the app and then synchronize based on that.
